keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,622 @@
+- *Extensibility*: Extending them is a verbose process, which leads to more sprawl in terms of test implementation.
+- *Performance*: They suffer from low performance due to the high number of pods created.  Network policy tests can take 30 minutes or longer.  The lack of completeness in positive controls, if fixed, could allow us to rapidly skip many tests destined for failure due to cluster health issues not related to network policy.
+- *Dynamic scale*: In addition to increasing the performance of these tests, we also should expand their ability to evaluate CNI's with highly dynamic, realistic workloads, outputting summary metrics.
+- *Documentation and Community*: The overall situation for these tests is that they are underdocumented and poorly understood by the community, and it's not clear how these tests are vetted when they are modified; this makes it difficult for CNI providers to compare and contrast compatibility and conformance to K8s standards for NetworkPolicys.
+- *Continuous Integration*: As part of this overall effort, once this test suite is more reliable and proven to be faster, running a basic verification of it in CI with some collection of CNI providers which could feed back into upstream K8s test results would be ideal, so that we know the NetworkPolicy test and specifications, as defined, are implemented/implementable correctly by at least some CNI provider.
+
+### Goals
+
+- Rearchitect the way we write and define CNI NetworkPolicy test verifications
+- Increase the visibility and quality of documentation available for network policies
+
+
+### Non-goals
+
+- Make tests specific to CNI providers
+- Change NetworkPolicy APIs
+ 
+### Related issues
+
+As an overall improvement, this KEP will help to address the solutions for several existing issues in upstream Kubernetes.  Some of these issues have been duct-taped upstream, but our overarching goal is to reduce the amount of work required to verify that any such issues have been properly addressed and accounted for in the documentation, testing, and semantic aspects of how the API for NetworkPolicy itself is defined.
+
+- https://github.com/kubernetes/kubernetes/issues/87857 (docs and understandability)
+- https://github.com/kubernetes/kubernetes/issues/87893 (holes in our test coverage matrix)
+- https://github.com/kubernetes/kubernetes/issues/85908 (failing tests, unclear semantics)
+- https://github.com/kubernetes/kubernetes/issues/86578 (needs e2e coverage)
+- https://github.com/kubernetes/kubernetes/issues/87709 (logging of netpol actions, will help describing states we reach) 
+- https://github.com/projectcalico/felix/issues/2032 (non-deterministic time frames for policy applications - addressable through published performance tests which measure time to policy implementation for several pods in several namespaces.)
+- https://github.com/projectcalico/felix/issues/2008 (need to test postStart pods in networkpolicy upstream.)
```

Comment:
It's not clear that this is a bug. It is not defined at what point in the pod lifecycle NetworkPolicy is expected to be in effect. This may require discussion and documentation updates which seem unrelated to the "make tests better" theme of this KEP.

Reply:
fixed pushing shortly